Neo4j.Driver/Neo4j.Driver/Internal/ConnectionValidator.cs

```diff
@@ -0,0 +1,116 @@
+using System;
+using System.Threading.Tasks;
+
+namespace Neo4j.Driver.Internal
+{
+    internal class ConnectionValidator
+    {
+        private readonly TimeSpan _connIdleTimeout;
+        private readonly TimeSpan _maxConnLifeTime;
+
+        public ConnectionValidator(TimeSpan connIdleTimeout, TimeSpan maxConnLifeTime)
+        {
+            _connIdleTimeout = connIdleTimeout;
+            _maxConnLifeTime = maxConnLifeTime;
+        }
+
+        public bool IsConnectionReusable(IPooledConnection connection)
+        {
+            if (!connection.IsOpen)
+            {
+                return false;
+            }
+
+            try
+            {
+                connection.ClearConnection();
+            }
+            catch
+            {
+                return false;
+            }
+
+            RestartIdleTimer(connection);
+
+            return true;
+        }
+
+        public async Task<bool> IsConnectionReusableAsync(IPooledConnection connection)
+        {
+            if (!connection.IsOpen)
+            {
+                return false;
+            }
+
+            try
+            {
+                await connection.ClearConnectionAsync().ConfigureAwait(false);
+            }
+            catch
+            {
+                return false;
+            }
+
+            RestartIdleTimer(connection);
+
+            return true;
+        }
+
+        private void RestartIdleTimer(IPooledConnection connection)
+        {
+            if (IsConnectionIdleDetectionEnabled())
+            {
+                connection.IdleTimer.Start();
```

Comment:
I think it is a bit strange that validator resets timers. Can this happen in the connection pool instead?
Same happens in `HasBeenIdleForTooLong`.